File: ch4/py/c-persistent-memory.py
from typing import Annotated, TypedDict

from langchain_core.messages import HumanMessage
from langchain_openai import ChatOpenAI
from langgraph.graph import StateGraph, START, END, add_messages
from langgraph.checkpoint.memory import MemorySaver


class State(TypedDict):
    messages: Annotated[list, add_messages]


builder = StateGraph(State)

model = ChatOpenAI()


def chatbot(state: State):
    answer = model.invoke(state["messages"])
    return {"messages": [answer]}


builder.add_node("chatbot", chatbot)
builder.add_edge(START, "chatbot")
builder.add_edge("chatbot", END)

# Add persistence with MemorySaver
graph = builder.compile(checkpointer=MemorySaver())

# Configure thread
thread1 = {"configurable": {"thread_id": "1"}}

# Run with persistence
result_1 = graph.invoke({"messages": [HumanMessage("hi, my name is Jack!")]}, thread1)
print(result_1)

result_2 = graph.invoke({"messages": [HumanMessage("what is my name?")]}, thread1)
print(result_2)

# Get state
print(graph.get_state(thread1))

File: ch4/js/c-persistent-memory.js
import {
  StateGraph,
  Annotation,
  messagesStateReducer,
  START,
} from "@langchain/langgraph";
import { ChatOpenAI } from "@langchain/openai";
import { MemorySaver } from "@langchain/langgraph";
import { HumanMessage } from "@langchain/core/messages";

const State = {
  messages: Annotation({
    reducer: messagesStateReducer,
    default: () => [],
  }),
};

let builder = new StateGraph(State);

const model = new ChatOpenAI();

async function chatbot(state) {
  const answer = await model.invoke(state.messages);
  return { messages: answer };
}

builder = builder.addNode("chatbot", chatbot);

builder = builder.addEdge(START, "chatbot").addEdge("chatbot", END);

// Add persistence
const graph = builder.compile({ checkpointer: new MemorySaver() });

// Configure thread
const thread1 = { configurable: { thread_id: "1" } };

// Run with persistence
const result_1 = await graph.invoke(
  {
    messages: [new HumanMessage("hi, my name is Jack!")],
  },
  thread1,
);
console.log(result_1);

const result_2 = await graph.invoke(
  {
    messages: [new HumanMessage("what is my name?")],
  },
  thread1,
);
console.log(result_2);

// Get state
await graph.getState(thread1);
